### DISCUSSION TITLE
isReadOnly true or false

### DISCUSSION THREAD

**Discussion by allenhwkim** [xyflow/xyflow#3254] [Ideas]
First, thanks for this product.

I think it's better to have  read-only mode, so that users can use reactflow as a part of a web page. e.g. select a reactflow node on the left-side, and users fill out some related information on the right side. 

What I want to see is  "reactflow is not affected by other keyboard events on an activity from other sections of a page". 
I believe that  there are many cases that reactflow is a part of a webpage. e.g., doing something with a diagram.

Currently `nokey` class is implemented to prevent other keyboard behaviour, but the issue is that we need to add `nokey` to every single html tags that accept keyboard events.

Here is my idea.

When `isReadOnly` is true, as a part of https://reactflow.dev/docs/api/react-flow-props/#interaction, it set these properties

* nodesDraggable : false (default true)
* nodesConnectable: false (default true)
* connectedOnClick : false (default true)
* elementsSelectable: true (default true)
* nodesDeletable: false (NEW)
* edgesDeletable: false (NEW)

By doing so, users are free to select a node or edge, but not affected by other keyboard interaction. Current behaviour is if user is abled to select a node or edge, they are also able to delete it by keyboard action(which may caused by other section keyboard event)

When `isReadOnly` is false, it set these properties

* nodesDraggable : true (default true)
* nodesConnectable: true (default true)
* connectedOnClick : true (default true)
* elementsSelectable: true (default true)
* nodesDeletable: true (NEW)
* edgesDeletable: true (NEW)

**Comment by bcakmakoglu** [comment]
I don't quite see why this would need to be added to the library when all of these props can be controlled by the user anyway? 😄 

Having another prop that disables other props is just inviting people to report "bugs" because they use mutually exclusive props with each other (which already is the case with some props tbh, so it would just exacerbate the issue).
What if I set `isReadOnly` to `true` but for some reason I set `nodesDraggable` also to `true` and now I am left wondering, why I can't drag nodes - nothing would prevent me from doing so.
With Generics there could be a typecheck to exclude some props if another is passed but that's making the Props interface way more complex than it needs to be imo.

So, personally, I don't really see much of a benefit in having this built-in when it's very simple to implement on the user-side of things.

Though I am in favor of adding props such as `nodesDeletable` and `edgesDeletable` (though you can set `deleteKeyCode` to `null` and disable the key altogether).

**Reply by allenhwkim** [reply]
I agree with you having another prop can cause some unnecessary bug reporting. However if documented correctly, false reporting can be avoided.

The reason I came up with this idea was that it was hard to find a solution to disable editing as not an expert.
Thus, this idea is to provide a simple solution to users for enabling or disabling editing.

Also, I think, if someone is just a diagram browsing user, not a publisher, then setting should be much simpler than coding many of properties, which they don't seem to care.

```jsx
    <ReactFlow
      nodes={nodes}
      edges={edges}
      readOnly={true}
    >
    </ReactFlow>
 ```
 vs.
```jsx
    <ReactFlow
      nodes={nodes}
      edges={edges}
      nodesDraggable={false}
      nodesConnectable={false}
      connectedOnClick={false}
      deleteKeyCode={null}
      /* more property setting to make it read-only, possibly*/
    >
    </ReactFlow>
```

**Reply by bcakmakoglu** [reply]
You could just abstract that though by wrapping `ReactFlow`, kind of making it even less necessary to have another prop to do this.

```tsx
const LockableFlow = ({ isReadOnly, ...props }) => {
	return <ReactFlow {...props} nodesDraggable={!isReadOnly} nodesConnectable={!isReadOnly} connectOnClick={!isReadOnly} />
}
```

And then just use it wherever you want without having to pass *all* the props to disable interactivity.

```tsx
const Flow = () => {
	const [nodes, setNodes, onNodesChange] = useNodesState(initialNodes)
	
	// ...
	
	return (
		<LockableFlow isReadOnly={isReadOnly} nodes={nodes} edges={edges) />
	)
}
```

Just my 2 cents though, maybe the devs have a different view on this 🤷‍♂️
